_posts/2017-11-21-Installing-Jenkins-with-PowerShell-DSC.md

```diff
@@ -0,0 +1,745 @@
+		Set-Content `
+			-Path "${ENV:ProgramFiles(x86)}\Jenkins\Jenkins.xml" `
+			-Value $NewConfig `
+			-Force
+		Write-Verbose -Verbose "Restarting Jenkins"
+	}
+	GetScript = {
+		$Config = Get-Content `
+			-Path "${ENV:ProgramFiles(x86)}\Jenkins\Jenkins.xml"
+		$Matches = @([regex]::matches($Config, "<arguments>[\s\S]*?<\/arguments>", 'IgnoreCase'))
+		$currentMatch = $Matches.Groups[1].Value
+		Return @{
+			'Result' = $currentMatch
+		}
+	}
+	TestScript = { 
+		$Config = Get-Content `
+			-Path "${ENV:ProgramFiles(x86)}\Jenkins\Jenkins.xml"
+		$Matches = @([regex]::matches($Config, "<arguments>[\s\S]*?<\/arguments>", 'IgnoreCase'))
+		$argString = "-Xrs -Xmx"+$Using:JenkinsXmx+"m -XX:MaxPermSize="+$Using:JenkinsMaxPermSize+"m -Djenkins.install.runSetupWizard=false -Dhudson.lifecycle=hudson.lifecycle.WindowsServiceLifecycle -jar `"%BASE%\jenkins.war`" --httpPort="+$Using:JenkinsPort+" --httpListenAddress=127.0.0.1 --webroot=`"%BASE%\war`""
+		$currentMatch = $Matches.Groups[1].Value
+		
+		Write-Verbose "Current service arguments: $currentMatch"
+		Write-Verbose "Should be service arguments: $argString"
+		
+		If ($argString -ne $currentMatch) {
+			# Jenkins port must be changed
+			Return $False
+		}
+		# Jenkins is already on correct port
+		Return $True
+	}
+	DependsOn = "[cChocoPackageInstaller]installJenkins"
+}
+```
+
+I admit. The script could be cleaner. Nevertheless it sets startup arguments in jenkins.xml file to have given memorylimit parameters and listen only localhost on the given port. We listen only localhost because I don't feel Jenkins should be open in any network. I'll later on show how to setup IIS to be a reverse proxy in front of the Jenkins. 
+
+
+## Initial authentication setup for Jenkins
+
+If we would startup the Jenkins at this moment it would not be ready to get api requests from us which we will need to install plugins. So what we need to do is to setup a initialization script to setup authentication. We can do it by making sure that there is groovy script under Jenkins/init.groovy.d/ folder. All the groovy scripts are run on that folder are run on the Jenkins startup. Yes. That is a bit scary. The actual script setup is easy one:
+
+
+```powershell
+File JenkinsAuthenticationSetup 
+{
+	DestinationPath = $JenkinsInitScriptPath
+	SourcePath = (Join-Path $InstallConfDirectory "solita_jenkins_security_realm.groovy")
+	Ensure = "Present"
+	Type = "File"
+	Checksum = "modifiedDate"
+	Force = $true
+	MatchSource = $true
+	DependsOn = "[cChocoPackageInstaller]installJenkins"
+}
+```
+
+So we put a groovy script there and it is set. That script checks the authorization strategy and sets it to be full control once logged in authorization strategy. It also creates a user with given username and password. The script file has placeholders for username and password so we just use again some a pretty ugly scripting to replace them with given parameters. 
```

Comment:
Vähän syö motivaatiota, kun blogin alussa on hehkutettu PowerShell DSC "Make scripting less complex" jne. , mutta tässä ja ylempänä haukutaan esimerkkiscriptejä rumaksi. Voisiko näitä siistiä, niin ei jää powershellistä huonoa mielikuvaa.